### PR TITLE
Han gatherv noncontiguous datatype fix

### LINKED ISSUE #12438: MTT MPI_Gatherv_c/MPI_Scatterv_c failure
### What version of Open MPI are you using? (e.g., v4.1.6, v5.0.1, git branch name and hash, etc.)

main branch

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Nightly tarball

### Please describe the system on which you are running

* Operating system/version: Amazon Linux 2
* Computer hardware: c6gn.16xlarge
* Network type: EFA

-----------------------------

## Details of the problem

https://mtt.open-mpi.org/index.php?limit=&wrap=&trial=&enable_drilldowns=&yaxis_scale=&xaxis_scale=&hide_subtitle=&split_graphs=&remote_go=&do_cookies=&phase=test_run&text_start_timestamp=2024-03-12+23%3A31%3A10+-+2024-03-26+23%3A31%3A10&text_platform_hardware=%5Eaarch64%24&show_platform_hardware=show&text_os_name=%5ELinux%24&show_os_name=show&text_mpi_name=%5Eompi-nightly-main%24&show_mpi_name=show&text_mpi_version=%5Emain-202403230241-984944d%24&show_mpi_version=show&text_suite_name=%5ETestBuild%3AIntelInstalled%24&show_suite_name=show&text_test_name=all&text_np=%5E256%24&show_np=show&text_full_command=&show_full_command=show&text_http_username=%5Eamazon%24&show_http_username=show&text_local_username=all&show_local_username=hide&text_platform_name=%5Eaws-amazonlinux-slurm-efa-installer%24&show_platform_name=show&click=Detail&phase=test_run&test_result=_run_f&text_compute_cluster_id=&text_os_version=&show_os_version=&text_platform_type=&show_platform_type=&text_submit_id=&text_hostname=&show_hostname=&text_mpi_get_id=&text_mpi_install_compiler_id=&text_mpi_install_configure_id=&text_mpi_install_id=&text_compiler_name=&show_compiler_name=&text_compiler_version=&show_compiler_version=&text_vpath_mode=&show_vpath_mode=&text_endian=&show_endian=&text_bitness=&show_bitness=&text_configure_arguments=&text_exit_value=&show_exit_value=&text_exit_signal=&show_exit_signal=&text_duration=&show_duration=&text_client_serial=&show_client_serial=&text_result_message=&text_result_stdout=&text_result_stderr=&text_environment=&text_description=&text_test_build_id=&text_test_build_compiler_id=&text_test_run_id=&text_test_name_id=&text_test_run_command_id=&text_test_suite_id=&text_performance_id=&text_launcher=&show_launcher=&text_resource_mgr=&show_resource_mgr=&text_network=&show_network=&text_parameters=&show_parameters=&lastgo=detail

This is caused by #12376

### PR DESCRIPTION
**Pull request by wenduwan**
This is a bugfix for derived datatypes with gaps. The reorder buffer should match the type signature and be large enough to hold gaps in addition to the actual entries.

Fixes #12438

### CODE REVIEW

**Review by devreal — APPROVED** [review]

**Comment by wenduwan** [comment]
Running AWS CI

File: ompi/mca/coll/han/coll_han_gatherv.c [hunk 1]
```diff
@@ -210,7 +207,9 @@ int mca_coll_han_gatherv_intra(const void *sbuf, int scount, struct ompi_datatyp
         }
 
         if (need_bounce_buf) {
-            bounce_buf = malloc(rdsize * total_up_rcounts);
+            ptrdiff_t rsize, rgap;
```
Comment (bosilca) [inline ompi/mca/coll/han/coll_han_gatherv.c:210]: Why do you care about this type ? This process is just an intermediary, for what he cares about he is just manipulating bytes. Of course if you go the MPI_BYTE route you cannot use `ompi_datatype_copy_content_same_ddt` instead you need to pack into a contiguous buffer with the correct size and at the correct index.

File: ompi/mca/coll/han/coll_han_scatterv.c [hunk 2]
```diff
@@ -144,8 +143,6 @@ int mca_coll_han_scatterv_intra(const void *sbuf, const int *scounts, const int
             low_scounts[low_peer] = scounts[w_peer];
         }
 
-        ompi_datatype_type_size(sdtype, &sdsize);
```
Comment (bosilca) [inline ompi/mca/coll/han/coll_han_scatterv.c:147]: Same comment as for gather except that you should use unpack to go from a packed buffer into the local type.

**Review by bosilca — COMMENTED** [review]
It is technically illegal in MPI to send one type and receive it into another type, with the exception of MPI_PACKED. I just mentioned it for the sake of the argument, I'm not saying we need to follow that rule. However, if we could document that this algorithm is not working in a heterogeneous environment, and make it disqualify itself it would be great.

File: ompi/mca/coll/han/coll_han_gatherv.c (hunk 1 above)
Comment (wenduwan) [reply ompi/mca/coll/han/coll_han_gatherv.c:210]: > It is technically illegal in MPI to send one type and receive it into another type, with the exception of MPI_PACKED.

Understood. I read that in the MPI standard and I think it is indeed the case for `MPI_Gatherv`. But inside ompi's implementation I figured that packing(on sender side) and unpacking(on receiver side) are already taken care of, which means we can indeed break away from the standard when calling `coll_gatherv`.

> this algorithm is not working in a heterogeneous environment
I need your help to understand this statement. What is a heterogeneous environment?

Comment (wenduwan) [reply ompi/mca/coll/han/coll_han_gatherv.c:210]: > This process is just an intermediary, for what he cares about he is just manipulating bytes. Of course if you go the MPI_BYTE route you cannot use ompi_datatype_copy_content_same_ddt instead you need to pack into a contiguous buffer with the correct size and at the correct index.

In this implementation we only use `MPI_BYTE` on node leaders(excluding the Root) since they don't know the type signature of the other processes - the root and node followers still use the true recv/send datatypes.

So overall the gatherv workflow looks like this:
- Node followers post send with the true send datatype. The data is packed automatically.
- The node leader posts recv with `MPI_BYTE` and stores node followers' packed data on a contiguous temp buffer. Then it posts send with `MPI_BYTE` from the same temp buffer - the data is already packed.
- The Root posts recv with the true recv datatype, and unpacks the data from the node leader into the recv buffer.

In other words, the node leader is transparent and only passes through the packed data from node followers to the Root.

Did I miss anything?

Comment (bosilca) [reply ompi/mca/coll/han/coll_han_gatherv.c:210]: It's kind of complicated. OMPI does the data conversion on the receiver, assuming the sender used a compatible datatype (aka. same type signature). So, the sender packs it's local representation into a packed representation, that once received will be unpacked using the sender configuration. And this is where the current scheme will break, because the `MPI_SOURCE` of the receive is not the initial source (where the data has been packed), but the intermediary node. This result in incorrect result after the unpacking.

Let me take an example, 3 nodes with 2 different endianness. A little endian, and B and C big endian. B will be out middleman here. When C receives the data it will come from B, which has the same endianness, so it incorrectly assumes there is nothing to be done. But the data comes in the format of A, which is of a different endianness, so it should have been converted. Long story short, in order to support heterogeneous environments, the middleman need to convert the data between the source and the local format, basically using the local datatype (with the correct count) in both, local and upstream, gatherv. While correct, this approach will be more costly, including for homogeneous environments using non-contiguous datatypes.

Comment (wenduwan) [reply ompi/mca/coll/han/coll_han_gatherv.c:210]: Thank you very much for the explanation! I wasn't aware of this issue.

> A little endian, and B and C big endian. B will be out middleman here. When C receives the data it will come from B, which has the same endianness, so it incorrectly assumes there is nothing to be done.

I have a couple follow up questions about this.

- What happens when B receives `MPI_BYTE` from A, which has a different endianness? I imagine the receiver, B, will figure out that A is different and "homogenize" the incoming data, so the data will be in B's endianness. In this case when B sends the data to A there shouldn't be additional conversion. I have no clue about this...
- A more detailed question - how does the receiver detect that the data is from a heterogenoeus sender?

Comment (wenduwan) [reply ompi/mca/coll/han/coll_han_gatherv.c:210]: @bosilca I also realized that this might not be a problem because the node leader and followers are on the same node so they should be equivalent....

**Comment by wenduwan** [comment]
@bosilca Thanks for your explanation wrt homogeneity. I have incorporated your feedback in the update.

Specifically, I added a flag in HAN to indicate homogeneity based on opal convertor's remote arch. I figured that only the scatterv algorithm is affected by this issue so I added a fallback logic. The gatherv algorithm shouldn't need this.

File: ompi/mca/coll/han/coll_han_scatterv.c [hunk 3]
```diff
@@ -94,6 +100,15 @@ int mca_coll_han_scatterv_intra(const void *sbuf, const int *scounts, const int
         return han_module->previous_scatterv(sbuf, scounts, displs, sdtype, rbuf, rcount, rdtype,
                                              root, comm, han_module->previous_scatterv_module);
     }
+    if (han_module->is_heterogeneous) {
+        printf("Not happy\n");
```
Comment (devreal) [inline ompi/mca/coll/han/coll_han_scatterv.c:104]: This should probably be taken out

File: ompi/mca/coll/han/coll_han_gatherv.c (hunk 1 above)
Comment (wenduwan) [reply ompi/mca/coll/han/coll_han_gatherv.c:210]: @bosilca I reckon that this is a problem for scatterv(node leader receives MPI_BYTE from Root which might have a different endianness) but not gatherv(node leader receives MPI_BYTE from node followers which have the same endianness). Therefore I added a homogeneity check in scatterv and bail in the heterogeneous case.

I also considered an alternative, i.e. when node leader receives MPI_BYTE(or similarly MPI_PACKED) from Root, it can swap bytes according to the convertor flags. However, after the discussion in slack I realized this only addresses the endianness conversion, and we also need to ensure the data type's byte representation be correctly translated. With everything considered, the heterogeneous case is rather rare in the field so I decided to be lazy and fall back to the other algorithm.

File: ompi/mca/coll/han/coll_han_scatterv.c (hunk 3 above)
Comment (wenduwan) [reply ompi/mca/coll/han/coll_han_scatterv.c]: Ooops fixed!

**Comment by jsquyres** [comment]
I should point out that in the OMPI docs, we clearly label the heterogeneous support in OMPI as broken:

Screenshot 2024-04-04 at 11 34 08 AM

This is in https://docs.open-mpi.org/en/v5.0.x/installing-open-mpi/configure-cli-options/misc.html.

**Comment by bosilca** [comment]
@ggouaillardet put a lot of efforts to make it work at some point. I haven't see anything going in OB1 that would break it, so I think that statement is a little to harsh and too generic.

**Comment by jsquyres** [comment]
Should the statement be removed, or softened?

@ggouaillardet Can you comment with some specifics, and/or post a PR to adjust/remove/whatever that statement in the docs?

File: ompi/mca/coll/han/coll_han_scatterv.c (hunk 2 above)
Comment (wenduwan) [reply ompi/mca/coll/han/coll_han_scatterv.c:147]: Currently inside ompi we need extra data size check in order to use MPI_PACKED - it is currently possible that the total byte size of `datatype size` x `count` can exceeded INT_MAX.

I think we should switch to explicit pack/unpack after large count support.

**Comment by wenduwan** [comment]
@bosilca Is there anything else that you would like to address? Unfortunately I do not have a hetero system to test the change so I can only disqualify the algorithm at best. Please let me know.

File: ompi/mca/coll/han/coll_han_scatterv.c (hunk 2 above)
Comment (wenduwan) [reply ompi/mca/coll/han/coll_han_scatterv.c:147]: @bosilca I took time to come up with possible optimizations following our discussion on Monday. I was trying to take advantage of the type map, but soon realized that this is not generally useful between node leader and follower.

For scatterv, the invariant is actually between Root and other processes:
> The type signature implied by sendcount[i], sendtype at the root must be equal to the type signature implied by recvcount, recvtype at MPI process i (however, the type maps may be different). This implies that the amount of data sent must be equal to the amount of data received, pairwise between each MPI process and the root. Distinct type maps between sender and receiver are still allowed.

When we focus on the node leader and its local neighbors, this information is not helpful, since they do not need to match in their respective `recvcount, recvtype`.

There is a simplification(rather than optimization) opportunity though, if and only if the node leader's `recvtype`'s type map size is the same than its local followers. In that case we don't need `MPI_BYTE` - we can safely use `recvtype` instead. This wouldn't bring better performance though.

Comment (bosilca) [reply ompi/mca/coll/han/coll_han_scatterv.c:147]: Because the typemap provided by the different processes must match when complemented with the count, we could use `MPI_Get_elements` to build any datatype used in the communication for as long as we know the size in bytes we are supposed to handle.

File: ompi/mca/coll/han/coll_han_topo.c [hunk 4]
```diff
@@ -92,12 +92,25 @@ mca_coll_han_topo_init(struct ompi_communicator_t *comm,
...
     int low_rank = ompi_comm_rank(low_comm);
     int low_size = ompi_comm_size(low_comm);
 
+    ompi_proc_t *up_proc = NULL;
+    ompi_proc_t *low_proc = NULL;
+
     int *topo = (int *)malloc(sizeof(int) * size * num_topo_level);
-    int is_imbalanced = 1;
-    int ranks_non_consecutive = 0;
+    int is_imbalanced = 1, ranks_non_consecutive = 0, is_heterogeneous = 0;
```
[the remark this answers is not in the document]
Comment (wenduwan) [reply ompi/mca/coll/han/coll_han_topo.c:102]: Thank you! I removed the low comm check as you correctly pointed out guaranteed node-local homogeneity.

**Review by bosilca — APPROVED** [review]
Please squash before merge.

**Comment by wenduwan** [comment]
@bosilca Thank you!

How would you like the commits to look like after squash?

**Comment by bosilca** [comment]
I understand there are two topics:

- allocate reorder buffer with gaps in gatherv and scatterv
- determine communicator homogeneity and disqualify hierarchical scatterv for heterogeneous communicators

**Comment by wenduwan** [comment]
Commit squashed and CI passed. Merging.
